### PR TITLE
Refactor P2P

### PR DESCRIPTION
**Pull request by itamarcps**
This major update marks a significant transformation of our P2P protocol's foundational structure. Previously based on boost::beast and websockets, we have migrated to using boost::asio and raw TCP sockets for the construction of the P2P network. This modification significantly enhances our performance capabilities and sets a new standard for our P2P network's operation.

Simultaneously, I have conducted thorough reviews of both sessions and managers to optimize their functionality. We've added read/write strands and writing queues for sessions to improve our communication efficiency. In addition, we've implemented several enhancements within the ManagerBase, ManagerDiscovery, and ManagerNormal classes to ensure operational fluidity.

Talking about fludity, for example, previously, our design made it possible for a session to wait for a complete write before writing again, leading to potential locks and delayed responses when broadcasting multiple times consecutively. To counteract this, we've must ensure that any operation requiring a mutex lock will only be held for a minimal amount of time. This substantially reduces the risk of bottlenecks, providing a smoother and more efficient experience and should be the rule for working on the P2P protocol, everything accesses it, it should be fluid.

Key changes introduced in this update include:

- NodeIDs are now a tuple of an address and port, as opposed to the previous 32-bit hash. We may reintroduce UUID at a later stage, alongside rdPoS validator signatures to enable efficient slashing.
- The Client and Server now share a singular Session class, streamlining the codebase and improving manageability.
- Clients are now instantiated by the ClientFactory class, improving system organization and design.
- The ClientFactory and Server now operate on an io_context, each running on 4 separate threads. This enhancement rectifies the previous issue of a dangling thread on the ClientSession.
- Additional minor changes include revisions in variable/function naming, improving readability and maintainability of the code.

Later today, I will be releasing  documentation that delves further into the operations of Sessions and Managers. This information will be accessible on the 'sparq-docs' repository under the 'development' branch.

### CODE REVIEW

File: src/net/p2p/server.cpp [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    this->do_accept();
+  }
+
+  void ServerListener::run() {
+    this->do_accept();
+  }
+
+  void ServerListener::stop() {
+    // Cancel is not available under Windows systems
+    #ifndef __MINGW32__
```
Comment (jcarraror) [inline src/net/p2p/server.cpp:32]: Why? We are using docker in Windows. Even if we're planning to have a full windows support for the SDK, Visual Studio does not use MINGW.

File: src/utils/utils.cpp [hunk 2]
```diff
@@ -19,7 +19,7 @@ void Utils::logToFile(std::string_view str) {
 }
 
 void Utils::logToDebug(std::string_view pfx, std::string_view func, std::string_view data) {
-  std::lock_guard lock(debug_mutex);
+  //  std::lock_guard lock(debug_mutex);
```
Comment (jcarraror) [inline src/utils/utils.cpp:22]: garbage?

File: tests/net/p2p/p2p.cpp [hunk 3]
```diff
@@ -358,135 +371,137 @@ namespace TP2P {
...
-			auto node3SessionsIDs = p2pNode3.getSessionsIDs();
-			auto node4SessionsIDs = p2pNode4.getSessionsIDs();
-			auto node5SessionsIDs = p2pNode5.getSessionsIDs();
-			auto node6SessionsIDs = p2pNode6.getSessionsIDs();
-			auto node7SessionsIDs = p2pNode7.getSessionsIDs();
-			auto node8SessionsIDs = p2pNode8.getSessionsIDs();
-			auto node9SessionsIDs = p2pNode9.getSessionsIDs();
-			auto node10SessionsIDs = p2pNode10.getSessionsIDs();
-      
+      auto node1SessionsIDs = p2pNode1.getSessionsIDs();
```
Comment (jcarraror) [inline tests/net/p2p/p2p.cpp:397]: Why auto instead of `std::vector<NodeID>`?

**Review by jcarraror — CHANGES_REQUESTED** [review]

File: tests/net/p2p/p2p.cpp (hunk 3 above)
Comment (itamarcps) [reply tests/net/p2p/p2p.cpp:397]: Doesn't really matter, the function naming already tells its type.

File: src/utils/utils.cpp (hunk 2 above)
Comment (itamarcps) [reply src/utils/utils.cpp]: Uncommented, it was a typo.

File: src/net/p2p/server.cpp (hunk 1 above)
Comment (itamarcps) [reply src/net/p2p/server.cpp]: Condition removed and error handling added for .cancel() and .close()

**Pull request merged**